kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k

```diff
@@ -71,6 +74,10 @@ module BITWISE-SIMPLIFICATION [symbolic]
        andBool X +Int Z ==Int maxUInt256
       [simplification, concrete(X, Z), comm]
 
+    rule [notMaxUInt5-bit-and-nested]:
+      notMaxUInt5 &Int ( ( notMaxUInt5 &Int X ) +Int Y ) => chop ( ( notMaxUInt5 &Int X ) +Int ( notMaxUInt5 &Int Y ) )
+      [simplification, concrete(Y)]
+
```

Comment:
This simplification seems pretty specific and the odd-one-out. Is it needed for the overall simplification (the test added here)? If not, maybe we can remove it, or add a test which specifically exercises it.

I also wonder if it can be made more general somehow....

Reply:
It feels specific, yes, but it comes from the compiler. This pattern has appeared both in Optimism and as part of the tests for @nwatson22's [PR](https://github.com/runtimeverification/kontrol/pull/254). I can make it more general:
but my intuition is that this is sufficient. We do already have specific lemmas concerning bounds on `notMaxUInt5 &Int lengthBytes(X)`, which also come from the compiler. The more general lemma would be:
```
    rule [mask-bit-and-nested]:
      M &Int ( ( M &Int X ) +Int Y ) => chop ( ( M &Int X ) +Int ( M &Int Y ) )
      requires #rangeUInt(256, M)
       andBool 0 <=Int X andBool 0 <=Int Y
       andBool M +Int 1 ==Int 2 ^Int log2Int(M +Int 1)
      [simplification, concrete(M, Y)]
```
covering the cases when the mask is not just the first 26 bytes, but it's not clear there's a use case for it.